xalia/Uia/UiaConnection.cs

```diff
@@ -641,6 +641,21 @@ private async Task TranslateMsaaEventTask(uint eventId, IntPtr hwnd, int idObjec
                 }
             }
 
+            string msaa_element_id;
+            if (idChild == 0)
+                msaa_element_id = $"msaa-hwnd-{hwnd}";
+            else
+                msaa_element_id = $"msaa-hwnd-{hwnd}-{idChild}";
+            if (msaa_elements_by_id.TryGetValue(msaa_element_id, out var msaa_element))

```

Comment:
Then, fetching the actual IAccessible is the only way to reliably correlate events?

Reply:
Yes, and even then you'd need some way to match the actual IAccessible the event was raised on to one in the existing tree. This is one of the big issues IAccessible2 solves with uniqueID.

Basically, MSAA seems to have been designed around the idea that each IAccessible would have a unique HWND associated with it. Children of this IAccessible that didn't have an HWND wouldn't get their own IAccessible interface, they'd have their values queried through the methods of their parent by passing their child ID as an argument.

This breaks with things like browsers, that have a tree full of IAccessible's all under one HWND. Once you have an IAccessible object from within this tree, its child ID is `CHILDID_SELF` and its HWND ends up being the same as the root object. To get around this, applications come up with unique child IDs that they'd use when raising an event on some object in the tree, this gets passed to the root IAccessible's `get_accChild`, and the IAccessible that raised the event is returned. Once you have it, there's no way to know what child ID was used to retrieve it from the root IAccessible, which also means you have no way to match it with another IAccessible in the tree.

Reply:
This last comment was meant in reply to your first, it got posted while I was still typing I guess :)

Reply:
I wonder if it'd be OK to keep this assumption, with the known limitation that if we can't get a unique ID from IAccessible2 or ISimpleDOM, we can't do anything with the events we get. I suspect that correlating a random IAccessible to a known one in the tree is possible, but that it would be less efficient than polling the attributes that we expect to change.

We'd still have to change the handling of EVENT_OBJECT_CREATE/EVENT_OBJECT_DESTROY. I guess for EVENT_OBJECT_CREATE we need to "dial out" to figure out the object's parent, and for  EVENT_OBJECT_DESTROY we need to look up the hwnd/child ID pair we get in the event.

Reply:
> My understanding is that ISimpleDOM is designed to allow us to get an id that we can use for this comparison, though I guess mshtml on Windows probably doesn't support ISimpleDOM.
> 
> Since we don't currently have a way to correlate objects except by child ID (at least for MsaaElement), maybe what we should do for now is print a warning when we get a non-zero child ID that we don't know about, while we do know about the HWND with CHILDID_SELF.
> 
> I guess this is also an issue with the EVENT_OBJECT_CREATE/EVENT_OBJECT_DESTROY case above, where we assume that if child ID isn't CHILDID_SELF, then CHILDID_SELF is the immediate parent. That probably doesn't hold for cases where there's a deeper tree within a single HWND.

For MSHTML, we can likely use the method described [here](https://learn.microsoft.com/en-us/previous-versions//ms528415(v=vs.85)?redirectedfrom=MSDN#accessing-the-internet-explorer-object-associated-with-an-accessible-object) to get an `IHTMLElement` interface, which might be enough to get us some kind of unique identifier. This should only be necessary on Windows though, for Wine's mshtml we have IAccessible2 from Gecko, so we have a uniqueID.

Reply:
I guess there also needs to be a distinction of: the HWND/child ID used for events, vs the IAccessible/HWND/child ID used to query the IAccessible, since those things will be different in this case.

Reply:
Well, maybe not. We don't actually compare HWND/child ID directly, we compare the internal unique ID string. That string can be based on IA2 or whatever while the overall structure still has the IAccessible with CHILD_ID self.

Reply:
Maybe, but I think I'd rather not dial out for events at all with the exception of EVENT_OBJECT_CREATE.

Reply:
Resolving because the eventual solution will be to derive matching unique id strings using IAccessible2.